packages/react-server/src/ReactFlightServer.js

```diff
@@ -138,39 +151,95 @@ function escapeStringValue(value: string): string {
 
 export function resolveModelToJSON(
   request: Request,
+  parent: {+[key: string | number]: ReactModel} | $ReadOnlyArray<ReactModel>,
+  key: string,
   value: ReactModel,
 ): ReactJSONValue {
-  if (typeof value === 'string') {
-    return escapeStringValue(value);
+  // Special Symbols
+  switch (value) {
+    case REACT_ELEMENT_TYPE:
+      return '$';
+    case REACT_SERVER_BLOCK_TYPE:
+      return '@';
+    case REACT_LAZY_TYPE:
+    case REACT_BLOCK_TYPE:
+      invariant(
+        false,
+        'React Blocks (and Lazy Components) are expected to be replaced by a ' +
+          'compiler on the server. Try configuring your compiler set up and avoid ' +
+          'using React.lazy inside of Blocks.',
+      );
   }
 
-  if (value === REACT_ELEMENT_TYPE) {
-    return '$';
+  if (parent[0] === REACT_SERVER_BLOCK_TYPE) {
+    // We're currently encoding part of a Block. Look up which key.
+    switch (key) {
+      case '1': {
+        // Module reference
+        let moduleReference: ModuleReference = (value: any);
+        try {
+          let moduleMetaData: ModuleMetaData = resolveModuleMetaData(
+            request.bundlerConfig,
+            moduleReference,
+          );
+          return (moduleMetaData: ReactJSONValue);
+        } catch (x) {
+          request.pendingChunks++;
+          let errorId = request.nextChunkId++;
+          emitErrorChunk(request, errorId, x);
+          return serializeIDRef(errorId);
+        }
+      }
+      case '2': {
+        // Query
+        let query: () => ReactModel = (value: any);
+        try {
+          // Attempt to resolve the query.
+          return query();
+        } catch (x) {
+          if (
+            typeof x === 'object' &&
+            x !== null &&
+            typeof x.then === 'function'
+          ) {
+            // Something suspended, we'll need to create a new segment and resolve it later.
+            request.pendingChunks++;
+            let newSegment = createSegment(request, query);
+            let ping = newSegment.ping;
+            x.then(ping, ping);
+            return serializeIDRef(newSegment.id);
+          } else {
+            // This query failed, encode the error as a separate row and reference that.
+            request.pendingChunks++;
+            let errorId = request.nextChunkId++;
+            emitErrorChunk(request, errorId, x);
+            return serializeIDRef(errorId);
+          }
+        }
+      }
+      default: {
+        invariant(
+          false,
+          'A server block should never encode any other slots. This is a bug in React.',
+        );
+      }
+    }
+  }
+
+  if (typeof value === 'string') {
+    return escapeStringValue(value);
   }
 
+  // Resolve server components.
   while (
     typeof value === 'object' &&
     value !== null &&
     value.$$typeof === REACT_ELEMENT_TYPE
   ) {
+    // TODO: Concatenate keys of parents onto children.
```

Comment:
What case is this about?

Reply:
Server components render away but their keys should still count.

```
return c ? <Foo key="a" /> : <Foo key="b" />;
```

```
function Foo() {
  return <div key="meh" />;
}
```

This renders `['$', 'div', 'meh', null]` regardless. But if this is an update request, it should clear the state because the key has changed.

It's not going to be perfectly compatible semantics since the component type should be part of the equation too but that's a bit too much and not necessary since it doesn't have state.

I think a good compromise would be to concatenate the key onto the child. Possibly with some escaping rules. Similar to what React.Children does.